### PR TITLE
Improve error messages for except* (B025, B029, B030, B904) #14791

### PR DESCRIPTION
**Pull request by smokyabdulrahman**
Improves error message for [except*](https://peps.python.org/pep-0654/) (Rules: B025, B029, B030, B904)

Example python snippet:
```python
try:
    a = 1
except* ValueError:
    a = 2
except* ValueError:
    a = 2

try:
    pass
except* ():
    pass

try:
    pass
except* 1:  # error
    pass

try:
    raise ValueError
except* ValueError:
    raise UserWarning
```
Error messages
Before:
```
$ ruff check --select=B foo.py
foo.py:6:9: B025 try-except block with duplicate exception `ValueError`
foo.py:11:1: B029 Using `except ():` with an empty tuple does not catch anything; add exceptions to handle
foo.py:16:9: B030 `except` handlers should only be exception classes or tuples of exception classes
foo.py:22:5: B904 Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling
Found 4 errors.
```
After:
```
$ ruff check --select=B foo.py
foo.py:6:9: B025 try-except* block with duplicate exception `ValueError`
foo.py:11:1: B029 Using `except* ():` with an empty tuple does not catch anything; add exceptions to handle
foo.py:16:9: B030 `except*` handlers should only be exception classes or tuples of exception classes
foo.py:22:5: B904 Within an `except*` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling
Found 4 errors.
```

Closes https://github.com/astral-sh/ruff/issues/14791

### CODE REVIEW

File: crates/ruff_linter/src/rules/flake8_bugbear/rules/duplicate_exceptions.rs [hunk 1]
```diff
@@ -42,13 +42,19 @@ use crate::registry::Rule;
...
     fn message(&self) -> String {
-        let DuplicateTryBlockException { name } = self;
-        format!("try-except block with duplicate exception `{name}`")
+        let DuplicateTryBlockException { name, is_star } = self;
+        let star = if *is_star {
+            "*".to_string()
+        } else {
+            String::new()
+        };
+        format!("try-except{star} block with duplicate exception `{name}`")
```
Comment (charliermarsh) [inline crates/ruff_linter/src/rules/flake8_bugbear/rules/duplicate_exceptions.rs:57]: For obscure reasons, let's do this like:
```rust
if is_star {
  format!("try-except* block with duplicate exception `{name}`")
} else {
  format!("try-except block with duplicate exception `{name}`")
}
```

(These messages get included as-is in the docs, so it's preferable to expand them like that sometimes.)

**Comment by charliermarsh** [comment]
If you're able, can you run `cargo test -p ruff_linter --lib` and then `cargo insta accept` to update the checked-in snapshots? You'll need to run `cargo install cargo-insta` first.

**Comment by smokyabdulrahman** [comment]
> If you're able, can you run `cargo test -p ruff_linter --lib` and then `cargo insta accept` to update the checked-in snapshots? You'll need to run `cargo install cargo-insta` first.

Yes, sure thing. I've ran the tests locally before committing. However, I didn't know how to accept my test modifications.

Thank you 👌 I will be doing that with in the next commit.

File: crates/ruff_linter/src/rules/flake8_bugbear/rules/duplicate_exceptions.rs (hunk 1 above)
Comment (smokyabdulrahman) [reply crates/ruff_linter/src/rules/flake8_bugbear/rules/duplicate_exceptions.rs:57]: sure thing 👍 
I will adopt this approach with all the rules.

**Comment by smokyabdulrahman** [comment]
BTW @charliermarsh 

Since I am going to fix a couple more rules also. (B025, B029, B030, B904)

Does it make sense to do the following change [in the statement parser](https://github.com/astral-sh/ruff/blob/2119dcab6ffb89613b0be4d40dc68b5a2f01526a/crates/ruff_python_parser/src/parser/statement.rs#L1549):
```rust
            ExceptHandler::ExceptHandler(ast::ExceptHandlerExceptHandler {
                type_,
                name,
                body: except_body,
                range: self.node_range(start),
                in_trystar: block_kind.is_star(), <---- `try` block kind
            }),
```

so that `in_trystar` is accessible in all the rules handlers.
or is that an overkill?

**Comment by charliermarsh** [comment]
@smokyabdulrahman -- I'd prefer not to change the parser / AST, if possible.

File: crates/ruff_python_ast_integration_tests/tests/visitor.rs [hunk 2]
```diff
@@ -247,7 +247,7 @@ impl Visitor<'_> for RecordVisitor {
         self.exit_node();
     }
 
-    fn visit_except_handler(&mut self, except_handler: &ExceptHandler) {
+    fn visit_except_handler(&mut self, except_handler: &ExceptHandler, _is_star: bool) {
```
Comment (MichaReiser) [inline crates/ruff_python_ast_integration_tests/tests/visitor.rs:250]: I'd prefer if we can keep the `Visitor` trait unchanged. 

You can access the `try` statement from inside the rules using:

```
let is_star = checker
                    .semantic()
                    .current_statement()
                    .as_try_stmt()
                    .is_some_and(|try_stmt| try_stmt.is_star);
```

Comment (smokyabdulrahman) [reply crates/ruff_python_ast_integration_tests/tests/visitor.rs:250]: Got it 👍 
I will be applying the changes.

Comment (MichaReiser) [reply crates/ruff_python_ast_integration_tests/tests/visitor.rs:250]: Thanks

Comment (smokyabdulrahman) [reply crates/ruff_python_ast_integration_tests/tests/visitor.rs]: Change applied.
It's much simpler now!

I should spend sometime reading the common structs' definitions next time. :+1:

**Review by MichaReiser — APPROVED** [review]
Thank you!

File: crates/ruff_linter/src/rules/flake8_bugbear/rules/duplicate_exceptions.rs [hunk 3]
```diff
@@ -209,14 +209,14 @@ pub(crate) fn duplicate_exceptions(checker: &mut Checker, handlers: &[ExceptHand
...
-        .as_try_stmt()
-        .is_some_and(|try_stmt| try_stmt.is_star);
     if checker.enabled(Rule::DuplicateTryBlockException) {
         for (name, exprs) in duplicates {
             for expr in exprs {
+                let is_star = checker
+                    .semantic()
+                    .current_statement()
+                    .as_try_stmt()
+                    .is_some_and(|try_stmt| try_stmt.is_star);
```
Comment (smokyabdulrahman) [inline crates/ruff_linter/src/rules/flake8_bugbear/rules/duplicate_exceptions.rs:219]: Just for learning purposes. @MichaReiser 
Why was this moved inside the `diagnostic-branch`?

I thought calculating it outside the loop once would be better.
